### PR TITLE
[None][fix] Restore DSpark disaggregated decoding accuracy

### PR DESCRIPTION
**Pull request by longlee0622**
## Summary

- Bootstrap DSpark's absolute decode position on generation-only disaggregated workers from the runtime target positions.
- Track the number of initialized entries in each DSpark rolling-window slot, mask unwritten circular-window entries, and reset/preserve that state across slot recycling and CUDA graph warmup.
- Exclude speculative scratch blocks before trimming stale or cache-reused SWA blocks for KV transfer.

## Root cause

Generation workers in disaggregated serving do not run prompt prefill. Their newly allocated DSpark slots therefore started at position zero and a start-position-only circular mask could expose rolling-window entries that had never been written. This degraded draft quality, including during the first 128 tokens.

Separately, SWA block lists may contain a speculative-capacity tail. Trimming the prompt prefix first at a page boundary could retain an uninitialized scratch block while discarding initialized prompt KV.

## Validation

- Pre-commit hooks: passed.
- Targeted GPU tests: `95 passed` covering DSpark CUDA graph capture/replay, position bootstrap, multi-accepted-token advancement, circular valid-length masking, slot lifecycle/reset, and V1/V2 SWA speculative-boundary transfer behavior.
- GB200 full GSM8K (1,319 samples): `96.21` flexible-extract / `96.29` strict-match.
  - DEP8 context + DEP8 generation
  - `max_batch_size=16`, concurrency 16
  - DSpark draft length 5 and position bootstrap enabled
  - CUDA graph enabled
  - KV cache manager V2 on both context and generation
  - CuteDSL MegaMoE backend

The end-to-end run uses the TensorRT-LLM 1.3.0rc24 container. Since current `main` has newer disaggregation configuration fields than rc24, the run overlays the branch's DSpark/model changes and the rc24-compatible equivalent of the transceiver change. The current-head transceiver implementation is covered by the targeted tests above.

## Dev Engineer Review

- DSpark disaggregated decoding now initializes absolute positions from runtime target positions.
- DSpark tracks valid rolling-window lengths and initialized slot state.
- Slot reset, context seeding, stale-slot cleanup, and CUDA graph warmup restore this state consistently.
- Attention masks unwritten rolling-window entries and supports optional `valid_len`.
- SWA KV transfer removes speculative scratch blocks before stale or cache-reused block trimming.
- Multi-beam requests reject speculative scratch blocks.
- The updated APIs propagate `valid_len` through `DSparkDraftModel`.
- Legacy masking remains unchanged when `valid_len` is not provided.
- No configuration or test-list files changed.
- No blocking correctness or consistency issue was identified.

## QA Engineer Review

- Added test coverage for partial rolling-window validity and draft-block indexing.
- Added test coverage for slot reset, context seeding, stale-slot cleanup, CUDA graph replay, and disaggregated generation state.
- Added test coverage for SWA block-cap trimming, packed beam tails, speculative-tail removal, multi-beam rejection, and D-Spark boundary handling.
- No test functions were removed.
- The PR reports 95 targeted GPU tests passing.
- No corresponding `tests/integration/test_lists/` coverage entries are provided for these test-code changes.
- Verdict: needs follow-up.

### CODE REVIEW

File: tensorrt_llm/_torch/disaggregation/transceiver.py [hunk 1]
```diff
@@ -293,11 +293,27 @@ def _create_kv_slice(self, req: LlmRequest) -> KVSlice:
...
+                # SWA block lists are ordered from the oldest valid prompt
+                # block to the speculative scratch tail. Remove the
+                # uninitialized scratch blocks before trimming stale prompt
+                # blocks; otherwise a boundary-crossing draft allocation can
+                # make us retain scratch and discard initialized prompt KV.
+                allocated_blocks = (
+                    req.prompt_len + self._kv_cache_manager.num_extra_kv_tokens + tpb - 1
+                ) // tpb
+                scratch_blocks = max(0, allocated_blocks - total_blocks)
+                if scratch_blocks > 0:
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/disaggregation/transceiver.py:321]: Moving `block_ids[:total_blocks]` into the `else` branch leaves the SWA path with no unconditional upper bound. When `num_extra_kv_tokens == 0` — i.e. any SWA request without speculative decoding — `scratch_blocks` is `0`, this whole block is skipped, and `block_ids` now reaches `_trim_packed_beam_block_ids` untrimmed. At `beam_width == 1`, `_split_packed_beam_block_ids` returns early and never consults `total_blocks`, so `beam0_block_ids[-expected_valid:]` selects from the tail and any block past `total_blocks` displaces initialized prompt KV — the same failure this PR fixes for the V2 list, on a path unrelated to speculative decoding. `_CacheReuseAdapterV1.get_block_ids` calls `get_batch_cache_indices` without `num_blocks_per_seq`, so the list length is whatever the request currently holds; the guard being removed here suggests `size > total_blocks` was already observed in practice. Could the cap be kept unconditionally, raised to `allocated_blocks`?

File: tensorrt_llm/_torch/disaggregation/transceiver.py [hunk 2]
```diff
@@ -293,11 +293,27 @@ def _create_kv_slice(self, req: LlmRequest) -> KVSlice:
...
+                    assert req.py_beam_width == 1, (
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/transceiver.py]: The beam check is an `assert`, so `python -O` drops it.

File: tensorrt_llm/_torch/disaggregation/transceiver.py [hunk 3]
```diff
@@ -293,11 +293,27 @@ def _create_kv_slice(self, req: LlmRequest) -> KVSlice:
...
+                        "speculative scratch blocks require beam_width == 1"
+                    )
+                    block_ids = (
+                        block_ids[:-scratch_blocks]
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/transceiver.py:325]: Removing blocks from the end only works while the missing ones stay at the front, which attention sinks would change.

File: tensorrt_llm/_torch/disaggregation/transceiver.py (hunk 1 above)
Comment (longlee0622) [reply tensorrt_llm/_torch/disaggregation/transceiver.py:321]: Good catch — you're right that removing the unconditional cap introduced a regression for oversized V1 SWA lists when num_extra_kv_tokens == 0.

Fixed in 465019b038. The SWA path now caps the beam-0 portion to allocated_blocks before stale-window trimming, while preserving valid packed beam-tail blocks. Speculative scratch blocks are still removed afterward.

I added regression coverage for:

an oversized non-speculative V1 SWA list, ensuring extra blocks cannot displace initialized prompt KV;
an oversized packed multi-beam list, ensuring the beam-0 cap preserves valid beam tails.
I also reran the DSpark/transceiver targeted suite, including the existing 1150/1151 speculative-boundary coverage: 97 passed. Thanks for catching this.

File: tensorrt_llm/_torch/disaggregation/transceiver.py (hunk 2 above)
Comment (longlee0622) [reply tensorrt_llm/_torch/disaggregation/transceiver.py]: Fixed in e5d6599d4e. I replaced the assertion with an explicit ValueError, so the beam-width validation is preserved under python -O. The corresponding regression test now checks the runtime exception.

File: tensorrt_llm/_torch/disaggregation/transceiver.py (hunk 3 above)
Comment (longlee0622) [reply tensorrt_llm/_torch/disaggregation/transceiver.py:325]: Added a comment in e5d6599d4e documenting this assumption. Current PyExecutor cache managers disable KV-cache token sinks, so stale SWA blocks form an evictable prefix and speculative scratch remains at the logical tail. If token sinks are enabled in the future, this path must use block-ordinal metadata to preserve the sink prefix rather than relying on flattened-list suffix trimming.

DeepSeek V4’s learned attention sink is virtual softmax-denominator state and does not occupy or reorder KV-cache blocks.
